### ISSUE TITLE
Rewrite Book 3 to allow for flipping faces of Cornell Light

### ISSUE THREAD

**Issue by trevordblack** [RayTracing/raytracing.github.io#662]
See #661

"
However! In Book 3, the flipface usage is reintroduced in order to make the rectangle-shaped light one-sided, to avoid bleed-through in the ceiling. Currently, Book 3 has no code for it. When the currently-merged version of Book 2 gets published, there will be no place for a reader to find the flip face methods. Perhaps a tiny remark, but something that occurred to me 😄
"

**Comment by trevordblack** [comment]
The options for this are:
1. Leave everything as is
2. Hard code into `emitted` that the back face illuminates. It works in this case, but makes the general case wrong.  _i.e_
```
    color emitted(const ray& r_in, const hit_record& rec, double u, double v, const point3& p)
    const override {
        if (!rec.front_face)
            return emit->value(u, v, p);
        return color(0,0,0);
    }
```
3. Subclass the light so that only the light in the ceiling points the other way
4. Create a subclass xz_rect whose outward_normal is down
5. Add a boolean to the rect class to specify which side of the rect is the front_face
6. Remove that subsection and try to live with the light bleed on the light, maybe move it closer to the ceiling

**Comment by trevordblack** [comment]
7. Rotate the light 90 degrees in any axis

**Comment by hollasch** [comment]
8. Take a look at #756. We have a lot of redundant classes just because we don't have a general quadrilateral primitive. That's all we need. There are two general approaches:
8a. Implement a canonical quadrilateral primitive (like <0,0,0> x <1,1,0>, facing in the +Z direction) and rely on standard transforms to position and orient these quadrilaterals.
8b. Implement a parameterized quadrilateral that takes position plus two U,V vectors to create a quadrilateral in any orientation.

Normally I'd like 8a, but our current transform support is quite limited, and I think that's appropriate. 8b will give us what we need with quite a bit more flexibility, is easy to understand and is easy to implement. It will deprecate `box` (though we will then just specify a box factory function using quadrilaterals), `xy_rect`, `xz_rect`, `yz_rect` and probably `flip_face`.

**Comment by trevordblack** [comment]
this is currently blocked by #292

**Comment by hollasch** [comment]
Fixed in ed119afc11dab9bd912d0c2872fbd780c0909f8c
